lorawan/src/subnet.rs

```diff
@@ -0,0 +1,389 @@
+pub fn is_local_devaddr(devaddr: u32, netid_list: Vec<u32>) -> bool {
+    let netid = parse_netid(devaddr);
+    is_local_netid(netid, netid_list)
+}
+
+pub fn devaddr_from_subnet(subnetaddr: u32, netid_list: Vec<u32>) -> u32 {
+    let netid = subnet_addr_to_netid(subnetaddr, netid_list.clone());
+    let (lower, _upper) = netid_addr_range(netid, netid_list);
+    devaddr(netid, subnetaddr - lower)
+}
+
+pub fn subnet_from_devaddr(devaddr: u32, netid_list: Vec<u32>) -> u32 {
+    let netid = parse_netid(devaddr);
+    let (lower, _upper) = netid_addr_range(netid, netid_list);
+    lower + nwk_addr(devaddr)
+}
+
+fn netid_class(netid: u32) -> u8 {
+    let netclass: u8 = (netid >> 21) as u8;
+    netclass
+}
+
+fn addr_len(netclass: u8) -> u32 {
+    let result: u32 = match netclass {
+        0 => 25,
+        1 => 24,
+        2 => 20,
+        3 => 17,
+        4 => 15,
+        5 => 13,
+        6 => 10,
+        7 => 7,
+        _ => 0,
+    };
+    result
+}
+
+#[allow(dead_code)]
+fn addr_bit_len(devaddr: u32) -> u32 {
+    let netid = parse_netid(devaddr);
+    addr_len(netid_class(netid))
+}
+
+fn id_len(netclass: u8) -> u32 {
+    let result: u32 = match netclass {
+        0 => 6,
+        1 => 6,
+        2 => 9,
+        3 => 11,
+        4 => 12,
+        5 => 13,
+        6 => 15,
+        7 => 17,
+        _ => 0,
+    };
+    result
+}
+
+fn subnet_addr_to_netid(subnetaddr: u32, netid_list: Vec<u32>) -> u32 {
+    for item in netid_list.clone() {
+        if subnet_addr_within_range(subnetaddr, item, netid_list.clone()) {
+            return item;
+        }
+    }
+    0
+}
+
+fn subnet_addr_within_range(subnetaddr: u32, netid: u32, netid_list: Vec<u32>) -> bool {
+    let (lower, upper) = netid_addr_range(netid, netid_list);
+    (subnetaddr >= lower) && (subnetaddr < upper)
+}
+
+fn var_net_class(netclass: u8) -> u32 {
+    let idlen = id_len(netclass);
+    let result: u32 = match netclass {
+        0 => 0,
+        1 => 0b10u32 << idlen,
+        2 => 0b110u32 << idlen,
+        3 => 0b1110u32 << idlen,
+        4 => 0b11110u32 << idlen,
+        5 => 0b111110u32 << idlen,
+        6 => 0b1111110u32 << idlen,
+        7 => 0b11111110u32 << idlen,
+        _ => 0,
+    };
+    result
```

Comment:
FYI: everything in rust is an expression, so you can write this as only the match:
```
fn var_net_class(netclass: u8) -> u32 {
    let idlen = id_len(netclass);
    match netclass {
         0 => 0,
         1 => 0b10u32 << idlen,
         2 => 0b110u32 << idlen,
         3 => 0b1110u32 << idlen,
         4 => 0b11110u32 << idlen,
         5 => 0b111110u32 << idlen,
         6 => 0b1111110u32 << idlen,
         7 => 0b11111110u32 << idlen,
         _ => 0,
    } // <- notice lack of semicolon. removing it lets the the expression (match block) value fall out
}
```